### PR TITLE
DM-40840: Refactor bright star subtraction framework

### PR DESCRIPTION
**Pull request by leeskelvin**

### CODE REVIEW

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+        within_radii &= bright_stars["radius_mm"] <= self.config.max_focal_plane_radius
+
+        bright_stars = bright_stars[within_bbox & within_radii]
+
+        self.log.info(
+            "Identified %i of %i reference catalog star%s that are in the field of view, are in the "
+            "range %s mag, and that have no neighboring stars in the range %s mag within %s arcsec.",
+            len(bright_stars),
+            len(ref_cat_full),
+            "" if len(ref_cat_full) == 1 else "s",
```
Comment (mrawls) [inline python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: This is cute, but I suggest dropping it since the whole rest of the sentence is assuming plural, e.g., "are in the field of view," and I don't think perfect grammar for the weird single star catalog situation is worth multiple lines of code.

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+        Returns
+        -------
+        bright_stars : `~astropy.table.Table`
+            Table of bright stars within the exposure.
+        """
+        bbox = input_exposure.getBBox()
+        wcs = input_exposure.getWcs()
+        detector = input_exposure.detector
+
+        within_region = ref_obj_loader.loadPixelBox(bbox, wcs, "phot_g_mean")
```
Comment (mrawls) [inline python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: Consider making the refcat column name configurable

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        # Generate unique footprint IDs for NEIGHBOR masking
+        input_MI.mask.addMaskPlane(NEIGHBOR_MASK_PLANE)
+        if isinstance(footprints, SourceCatalog):
+            footprints = footprintsToNumpy(footprints, bbox, asBool=False)
+
+        # Establish pixel-to-boresight-pseudopixel transform
+        pixel_scale = input_exposure.wcs.getPixelScale(bbox.getCenter()).asArcseconds() * arcseconds
+        pixels_to_boresight_pseudopixels = input_exposure.detector.getTransform(PIXELS, FIELD_ANGLE).then(
+            makeTransform(AffineTransform.makeScaling(1 / pixel_scale.asRadians()))
+        )
```
Comment (mrawls) [inline python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: ... what just happened

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        stamp_radius = floor(Extent2D(*self.config.stamp_size) / 2)
+        stamp_bbox = Box2I(Point2I(0, 0), Extent2I(1, 1)).dilatedBy(stamp_radius)
+        stamp_radius_padded = floor((Extent2D(*self.config.stamp_size) * 1.42) / 2)
+        stamp_bbox_padded = Box2I(Point2I(0, 0), Extent2I(1, 1)).dilatedBy(stamp_radius_padded)
+
+        stamps = []
+        for bright_star in bright_stars:
+            pix_coord = Point2D(bright_star["pixel_x"], bright_star["pixel_y"])
+
+            # Set NEIGHBOR mask plane for all sources except the current one
```
Comment (mrawls) [inline python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: I guess this implies that the neighbor criteria have to be looser than the target criteria. What if weird configs were set so that isn't the case?

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+            # Warp the image and mask to the stamp frame
+            stamp_MI = MaskedImageF(stamp_bbox_padded)
+            warpImage(stamp_MI, input_MI, pixels_to_stamp_frame, warp_control)
+            stamp_MI = stamp_MI[stamp_bbox]
+
+            # Check mask coverage, update metadata
+            bad_bit_mask = stamp_MI.mask.getPlaneBitMask(self.config.bad_mask_planes)
+            good = (stamp_MI.mask.array & bad_bit_mask) == 0
+            good_frac = np.sum(good) / stamp_MI.mask.array.size
+            if good_frac < self.config.min_area_fraction:
```
Comment (mrawls) [inline python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: And if not... fall over?

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+    LoadReferenceObjectsConfig,
+    ReferenceObjectLoader,
+    WarpedPsf,
+)
+from lsst.pex.config import ChoiceField, ConfigField, Field, ListField
+from lsst.pipe.base import PipelineTask, PipelineTaskConfig, PipelineTaskConnections, Struct
+from lsst.pipe.base.connectionTypes import Input, Output, PrerequisiteInput
+from lsst.utils.timer import timeMethod
+
+NEIGHBOR_MASK_PLANE = "NEIGHBOR"
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: Here's a snippet from my testing just now:
```
ipdb> input_MI.mask.getMaskPlaneDict()
{'BAD': 0, 'CR': 3, 'CROSSTALK': 11, 'DETECTED': 5, 'DETECTED_NEGATIVE': 6, 'EDGE': 4, 'INTRP': 2, 'ITL_DIP': 12, 'NOT_DEBLENDED': 13, 'NO_DATA': 8, 'PARTLY_VIGNETTED': 14, 'SAT': 1, 'SPIKE': 15, 'STREAK': 10, 'SUSPECT': 7, 'UNMASKEDNAN': 16, 'VIGNETTED': 9}

ipdb> input_MI.mask.getMaskPlaneDict()["NEIGHBOR"]
*** KeyError: 'NEIGHBOR'

ipdb> input_MI.mask.getMaskPlaneDict().keys()
dict_keys(['BAD', 'CR', 'CROSSTALK', 'DETECTED', 'DETECTED_NEGATIVE', 'EDGE', 'INTRP', 'ITL_DIP', 'NOT_DEBLENDED', 'NO_DATA', 'PARTLY_VIGNETTED', 'SAT', 'SPIKE', 'STREAK', 'SUSPECT', 'UNMASKEDNAN', 'VIGNETTED'])

ipdb> input_MI.mask.addMaskPlane(NEIGHBOR_MASK_PLANE)
17

ipdb> input_MI.mask.getMaskPlaneDict().keys()
dict_keys(['BAD', 'CR', 'CROSSTALK', 'DETECTED', 'DETECTED_NEGATIVE', 'EDGE', 'INTRP', 'ITL_DIP', 'NEIGHBOR', 'NOT_DEBLENDED', 'NO_DATA', 'PARTLY_VIGNETTED', 'SAT', 'SPIKE', 'STREAK', 'SUSPECT', 'UNMASKEDNAN', 'VIGNETTED'])

ipdb> input_MI.mask.addMaskPlane(NEIGHBOR_MASK_PLANE)
17

ipdb> input_MI.mask.getMaskPlaneDict().keys()
dict_keys(['BAD', 'CR', 'CROSSTALK', 'DETECTED', 'DETECTED_NEGATIVE', 'EDGE', 'INTRP', 'ITL_DIP', 'NEIGHBOR', 'NOT_DEBLENDED', 'NO_DATA', 'PARTLY_VIGNETTED', 'SAT', 'SPIKE', 'STREAK', 'SUSPECT', 'UNMASKEDNAN', 'VIGNETTED'])
```

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+    )
+    exclude_arcsec_radius = Field[float](
+        doc="Stars with a star in the range ``exclude_mag_range`` mag in ``exclude_arcsec_radius`` are not "
+        "used.",
+        default=5,
+    )
+    exclude_mag_range = ListField[float](
+        doc="Stars with a star in the range ``exclude_mag_range`` mag in ``exclude_arcsec_radius`` are not "
+        "used.",
+        default=[0, 20],
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: Yes, thank you. Reworded to this which is hopefully clearer:
```
doc="No postage stamp will be generated for stars with a neighboring star in the range "
"``exclude_mag_range`` mag within ``exclude_arcsec_radius`` arcseconds.",
```

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py (hunk 3 above)
Comment (leeskelvin) [reply python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: Kind of yes. We’re setting up a transform that goes from pixel coordinates -> focal-plane positions in radians -> a uniform pseudopixel frame where each pixel has the same scale as the origin.

We can’t just use raw pixels because the pixel scale varies across the exposure. To warp and stack the cutouts into a single consistent stamp, we need a common reference coordinate system for all stars.

File: tests/test_brightStarSubtraction.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        # Background coefficients
+        sigma = 60.0  # noise
+        pedestal = 3210.1
+        coef_x = 1e-2
+        coef_y = 2e-2
+        coef_x2 = 1e-5
+        coef_xy = 2e-5
+        coef_y2 = 3e-5
+
+        # Make an input exposure
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply tests/test_brightStarSubtraction.py]: There are a few things that come close (I can't think of where they live right now), but I did have a rather specific setup that I wanted to initialize with here so I don't mind setting things up myself.

File: tests/test_brightStarSubtraction.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            "CR",
+            "CROSSTALK",
+            "EDGE",
+            "NO_DATA",
+            "SAT",
+            "SUSPECT",
+            "UNMASKEDNAN",
+            "NEIGHBOR",
+        ]:
+            _ = self.exposure.mask.addMaskPlane(mask_plane)
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply tests/test_brightStarSubtraction.py]: I'm not able to reproduce it when I attempt to add the same mask plane multiple times on an exposure/masked image. Can you point me to where/how this triggers?

File: python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+            A set of postage stamp cutouts, each centered on a bright star.
+        """
+        bright_stars = self._get_bright_stars(ref_obj_loader, input_exposure)
+
+        bright_star_stamps = self._get_bright_star_stamps(
+            input_exposure,
+            input_background,
+            input_source_catalog,
+            bright_stars,
+            visit=int(data_id["visit"]),
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/pipe/tasks/brightStarSubtraction/brightStarCutout.py]: Ah, I wasn't checking for the extra `.info` namespace. Yes, that seems to do the trick:
```
> self.exposure.info.setVisitInfo(VisitInfo(id=12345))
> self.exposure.visitInfo.getId()
12345
```
Cheers.

File: tests/test_brightStarSubtraction.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mrawls) [reply tests/test_brightStarSubtraction.py]: Looks like some guy named Lee ran into it a few years back, haha https://rubin-obs.slack.com/archives/C08CANY4B6H/p1624367768044800 

Here's a more recent example of when James hit this problem https://rubin-obs.slack.com/archives/C07TD4VUJ3X/p1764193741629889

If you're convinced this can't happen here / doesn't apply, that is also OK.

File: tests/test_brightStarSubtraction.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        brightStarCutoutTask = BrightStarCutoutTask(config=brightStarCutoutConfig)
+        self.bright_star_stamps = brightStarCutoutTask._get_bright_star_stamps(
+            input_exposure=self.exposure,
+            input_background=None,
+            footprints=self.footprints,
+            bright_stars=self.bright_stars,
+            visit=12345,
+        )
+
+    def test_brightStarCutout(self):
```
[the remark this answers is not in the document]
Comment (mrawls) [reply tests/test_brightStarSubtraction.py]: This isn't a blocker, but yes, I was trying to come up with some edge cases that might be good to test are handled in the way you would prefer.
